### DISCUSSION TITLE
Improved offline experience

### DISCUSSION THREAD

**Discussion by vishnukvmd** [ente/ente#10530] [Enhancements]
Desktop app should be able to work completely offline by rendering cached thumbnails and watched / exported items, when available.

We should gracefully handle network failures and surface that information as a toast / sticky header at the top.

**Comment by benjuade** [comment]
Same said for the mobile app. Impossible to delete a photo if they device is not connected to internet.

**Comment by adrienbar** [comment]
Being able to render pre-downloaded photos in ente desktop app would make the experience awesome @vishnukvmd !

Trying to render from local storage (e.g. the export folder !) before trying to fetch from server would also greatly improve the experience under poor/medium network conditions (removing the delay between the blurry version of the image and its full size version fetched from the server, and ideally making the whole gallery scrolling smoother thanks to locally stored thumbnails). 

 'Basically' (I know really is always more complexe ;-) ):  
 instead of 
 - render from cache if available
 - fetch from server
 
 It could be:
 - render from cache if available
 **- render from local export if available**
 - fetch from server 
 
 
And thank you for all the work with the app it is really great !

**Reply by adrienbar** [reply]
I tested the idea of reading from export on a local branch and this significantly improved the experience for me.

I wanted to make a PR but it seems I do not have permissions to push my branch to the remote repo.
So here is the git patch file if that can help (based on https://github.com/ente/ente/pull/13043)
[perf(Desktop)-read-images-from-export-folder.patch](https://github.com/user-attachments/files/32445032/perf.Desktop.-read-images-from-export-folder.patch) (`git apply "perf(Desktop)-read-images-from-export-folder.patch"`)

 **In scope:** The following items are read from the local export folder if they exist, else fetched from server (or un fetch cache):
 -  Images
 _NB: The images formats that can't be read natively by the chromium (electron app) - like heic - despite being read from local storage still need to go through trans-coding to jpeg, which can make the experience laggy when willing to navigate fast in the gallery.  I am thinking about storing the transcoded jpeg (or webp/ avih) versions locally as well but that is not part of the patch here._ 
 
 **Out of scope:** The following items are still always fetched from server (unless in fetch cache): 
 - Thumbnails
 - Video and live photos 
 
 **Tests made:** (in addition to the unit tests included)
 - built the linux x64 version of the app and run it
 - logged into an account with a small library ~90 photos jpg and heic (self hosted instance)
 - enabled export and waited for it to complete
 - navigated the gallery **online**: when navigating fast much less lag to display the 'none blurry' version of image
 - navigated the gallery **offline**: all photos rendered 'none blurry' 
 
 **Implementation notes**:
 The idea was to add a 'local' counterpart of the DownloadManager that provides a RenderableSourceURLs given a EnteFile, and to add a try call to it before calling the DownloadManager. The export-reader.ts consumes from the exporter (to get the localPath given an EnteFile) and helps to implement the LocalExportReader interface that is the counter part of the FileViewerDataSourceDownloadManager interface. Due to the import structure _app.tsx , data-source.ts and data-source-core.ts all need to have imports updated
 
 
Overall this adjustment saves bandwidth (fetch) and cpu (decrypt) without requiring extra storage (given that export is enabled), so it would be really nice to include it in the standard :-) @vishnukvmd  

This is one of my first contribution to a public repo, I hope it meets the standards :-)

**Comment by AswinAsok** [comment]
Hi @adrienbar,

I respect on the effort you've put in and this does look like something we can add to the app, so I've added it to my roadmap.

We have an entire desktop revamp planned for this year, and I'll try to include this while we're shipping it.

I will keep you updated in this thread.
